File: workflows/java/sdk/order-processor/src/main/java/io/dapr/quickstarts/workflows/activities/ProcessPaymentActivity.java
package io.dapr.quickstarts.workflows.activities;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import io.dapr.quickstarts.workflows.models.PaymentRequest;
import io.dapr.workflows.runtime.WorkflowActivity;
import io.dapr.workflows.runtime.WorkflowActivityContext;

public class ProcessPaymentActivity implements WorkflowActivity {
  private static Logger logger = LoggerFactory.getLogger(ProcessPaymentActivity.class);

  @Override
  public Object run(WorkflowActivityContext ctx) {
    PaymentRequest req = ctx.getInput(PaymentRequest.class);
    logger.info("Processing payment: {} for {} {} at ${}",
        req.getRequestId(), req.getQuantity(), req.getItemBeingPurchased(), req.getAmount());

    // Simulate slow processing
    try {
      Thread.sleep(7 * 1000);
    } catch (InterruptedException e) {
    }
    logger.info("Payment for request ID '{}' processed successfully", req.getRequestId());

    return true;
  }

}

File: workflows/csharp/sdk/order-processor/Activities/ProcessPaymentActivity.cs
namespace WorkflowConsoleApp.Activities;

using System.Threading.Tasks;
using Dapr.Client;
using Dapr.Workflow;
using Microsoft.Extensions.Logging;
using Models;
using System;

internal sealed partial class ProcessPaymentActivity(ILogger<ProcessPaymentActivity> logger) : WorkflowActivity<PaymentRequest, object?>
{
    public override async Task<object?> RunAsync(WorkflowActivityContext context, PaymentRequest req)
    {
        LogPaymentProcessing(logger, req.RequestId, req.Amount, req.ItemBeingPurchased, req.Currency);

        // Simulate slow processing
        await Task.Delay(TimeSpan.FromSeconds(7));

        LogSuccessfulPayment(logger, req.RequestId);
        return null;
    }

    [LoggerMessage(LogLevel.Information, "Processing payment: request ID '{requestId}' for {amount} {itemBeingPurchased} at ${currency}")]
    static partial void LogPaymentProcessing(ILogger logger, string requestId, int amount, string itemBeingPurchased, double currency);

    [LoggerMessage(LogLevel.Information, "Payment for request ID '{requestId}' processed successfully")]
    static partial void LogSuccessfulPayment(ILogger logger, string requestId);
}